In [2]:
import pandas as pd
import statsmodels.formula.api as smf
import numpy as np
import statsmodels.formula.api as smf

In [3]:
df_train = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTrain.csv')
df_test = pd.read_csv('https://raw.githubusercontent.com/amandeep0/IS451/main/data/ClimateChangeTest.csv')

(a) Data is split for training (<=2006) & testing (after 2006).

In [4]:
df_train.head()

Year  Month    MEI     CO2      CH4      N2O    CFC11    CFC12        TSI  \
0  1983      5  2.556  345.96  1638.59  303.677  191.324  350.113  1366.1024   
1  1983      6  2.167  345.52  1633.71  303.746  192.057  351.848  1366.1208   
2  1983      7  1.741  344.15  1633.22  303.795  192.818  353.725  1366.2850   
3  1983      8  1.130  342.25  1631.35  303.839  193.602  355.633  1366.4202   
4  1983      9  0.428  340.17  1648.40  303.901  194.392  357.465  1366.2335   

   Aerosols   Temp  
0    0.0863  0.109  
1    0.0794  0.118  
2    0.0731  0.137  
3    0.0673  0.176  
4    0.0619  0.149

In [6]:
df_linReg = smf.ols("Temp ~ MEI + CO2 + CH4 + N2O + CFC11 + CFC12 + TSI + Aerosols", data = df_train)
df_linReg_results = df_linReg.fit()
print(df_linReg_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     103.6
Date:                Mon, 07 Oct 2024   Prob (F-statistic):           1.94e-78
Time:                        22:32:23   Log-Likelihood:                 280.10
No. Observations:                 284   AIC:                            -542.2
Df Residuals:                     275   BIC:                            -509.4
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -124.5943     19.887     -6.265      0.0

(i) Linear Regression equation from model is:

Temp(i) = -124.59 + 0.0642*MEI(i) + 0.0065*CO2(i) + 0.0001*CH4(i) - 0.0165*N2O(i) - 0.0066*CFC11(i) + 0.0038*CFC12(i) + 0.0931*TSI(i) - 1.5376*Aerosols(i) + error(i)

(ii) R-squared value is 0.751. This measures the fit and is better when higher. It tends to increase with more variables. 75.1% R-square looks large enough implying that model is a good fit to the data.

A lower p-value means more chances that the variable has an effect. Let's take a benchmark of around 5%.
*   MEI, CFC11, CFC12, TSI, Aerosols - p-values are almost 0 for them implying these are very significant variables.
*   CO2 has 0.5% p-value which is again very low implying that CO2 is a significant variable.
*   N2O has 5.5% p-value meaning this is somewhat significant.
*   CH4 has a 81% p-value meaning it is not really significant.

(iii) Since our model is a "multiple" linear regression model, there might be variables correlating to each resulting in multicollinearity. Rather than looking at individual variables, model is looking at an overall picture where different variables may interact with each other, leading to negative co-efficients at places where individual variable may actually give a positive co-efficient. So, N2O & CFC11 may be correlated with other variables leading to negative coefficient estimates.

In [9]:
correlation_matrix = df_train.corr()
print("\nCorrelation matrix:")
print(correlation_matrix)


Correlation matrix:
              Year     Month       MEI       CO2       CH4       N2O  \
Year      1.000000 -0.027942 -0.036988  0.982749  0.915659  0.993845   
Month    -0.027942  1.000000  0.000885 -0.106732  0.018569  0.013632   
MEI      -0.036988  0.000885  1.000000 -0.041147 -0.033419 -0.050820   
CO2       0.982749 -0.106732 -0.041147  1.000000  0.877280  0.976720   
CH4       0.915659  0.018569 -0.033419  0.877280  1.000000  0.899839   
N2O       0.993845  0.013632 -0.050820  0.976720  0.899839  1.000000   
CFC11     0.569106 -0.013111  0.069000  0.514060  0.779904  0.522477   
CFC12     0.897012  0.000675  0.008286  0.852690  0.963616  0.867931   
TSI       0.170302 -0.034606 -0.154492  0.177429  0.245528  0.199757   
Aerosols -0.345247  0.014890  0.340238 -0.356155 -0.267809 -0.337055   
Temp      0.786797 -0.099857  0.172471  0.788529  0.703255  0.778639   

             CFC11     CFC12       TSI  Aerosols      Temp  
Year      0.569106  0.897012  0.170302 -0.345247  0.7

(iv) Both N2O & CFC11 are highly positive correlated with below variables :-

*   N2O --> CO2, CH4, CFC12
*   CFC11 --> CH4, CFC12

In [11]:
df_linReg2 = smf.ols("Temp ~ MEI + N2O + TSI + Aerosols", data = df_train)
df_linReg2_results = df_linReg2.fit()
print(df_linReg2_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.726
Model:                            OLS   Adj. R-squared:                  0.722
Method:                 Least Squares   F-statistic:                     184.9
Date:                Mon, 07 Oct 2024   Prob (F-statistic):           3.52e-77
Time:                        22:59:37   Log-Likelihood:                 266.64
No. Observations:                 284   AIC:                            -523.3
Df Residuals:                     279   BIC:                            -505.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -116.2269     20.223     -5.747      0.0

(b)

(i) N2O in this model is positive which seems plausible given that it is a greenhouse gas and hence, should increase temperature with increase in N2O. In previous model, N2O had high correlations with CO2, CH4, CFC12 which are no more present in this model. With fewer variables, the model now captures clearer relationship.

(ii) R-square is slightly higher for previous model; however, the model included lot more variables compared to second one. All of the variables in second model are highly significant, unlike first one, which had some unnecessary overlapping variables. Hence, second model is a better choice.

In [12]:
df_linReg2_results.predict(df_test)

0     0.504243
1     0.475001
2     0.450627
3     0.439461
4     0.450831
5     0.418853
6     0.422773
7     0.416541
8     0.370516
9     0.379791
10    0.376987
11    0.383793
12    0.398897
13    0.376185
14    0.355217
15    0.404700
16    0.444619
17    0.472767
18    0.462425
19    0.444292
20    0.422654
21    0.421001
22    0.438793
23    0.439419
dtype: float64

In [15]:
SSE = np.sum((df_linReg2_results.predict(df_test)-df_test['Temp'])**2)
SSE

0.2948967082567806

In [17]:
SST = np.sum((df_train['Temp'].mean()-df_test['Temp'])**2)
SST

0.5860188540964095

In [18]:
R2 = 1-(SSE/SST)
R2

0.4967794872206187

An R-quare of approximately 50% on test set means that it somewhat predicts temperature, but does not fully capture the complexities of data. There might be other relevant factors we may want to add to better predict temperature. However, too high R-square value may also indicate over-fitting which is not there in our model.

In climate studies, where variability is high and many external factors can play a role, 50% R-square can be reasonable.